source/codegen/metadata_validation.py

```diff
@@ -0,0 +1,287 @@
+from typing import Any, Dict, Iterable, List, Set
+from schema import Schema, And, Optional, Or, Use
+import common_helpers
+
+
+# Rules that can be suppressed
+class RULES:
+    # In general, arrays should have sizes passed in to the underlying C API.
+    ARRAY_PARAMETER_NEEDS_SIZE = "ARRAY_PARAMETER_NEEDS_SIZE"
+    # The gRPC layer can determine the size of input arrays, so we should not require
+    # callers to pass in the size. We should not add any new suppressions of this type.
+    INPUT_ARRAY_SHOULD_NOT_HAVE_PASSED_IN_SIZE = "INPUT_ARRAY_SHOULD_NOT_HAVE_PASSED_IN_SIZE"
+    # In general, enums shouldn't have duplicate values. This is helpful for catching typos
+    # although there are a few enums that have legitimate duplicates.
+    ENUMS_SHOULD_NOT_HAVE_DUPLICATE_VALUES = "ENUMS_SHOULD_NOT_HAVE_DUPLICATE_VALUES"
+
+
+def validate_metadata(metadata: dict):
+    try:
+        for function_name in metadata['functions']:
+            validate_function(function_name, metadata)
+        for attribute_group in common_helpers.get_attribute_groups(metadata):
+            for attribute_id in attribute_group.attributes:
+                validate_attribute(attribute_group.attributes[attribute_id], metadata)
+        function_enums = get_function_enums(metadata['functions'])
+        attribute_enums = get_attribute_enums(metadata)
+        used_enums = function_enums.union(attribute_enums)
```

Comment:
note: Fun fact - we only generate enums from `enums.py` that are actually used in a function or attribute. So pass those in to `validate_enum` so we can know how much to check them.

Reply:
Were there a lot of violations in unused enums? I'd run into that fun fact before and been annoyed by it. I'm not sure what to do about it (there's also some value in making fewer changes to nimi-python-derived metadata; so I'm not sure we should torch the unused ones). We could make "allow-unused-enums" be optional or something.

Reply:
(not part of this PR, just curious)

Reply:
I'm not sure - because validation stops as soon as it finds a problem (this is something that could be improved), all I know is it came up at least once in a few different drivers.